% Created 2019-03-17 Sun 21:28
% Intended LaTeX compiler: pdflatex
\documentclass[11pt]{article}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{graphicx}
\usepackage{grffile}
\usepackage{longtable}
\usepackage{wrapfig}
\usepackage{rotating}
\usepackage[normalem]{ulem}
\usepackage{amsmath}
\usepackage{textcomp}
\usepackage{amssymb}
\usepackage{capt-of}
\usepackage{hyperref}
\usepackage{listings}
\input{../lit-header}
\date{\today}
\title{Crossover in ROPER 2}
\hypersetup{
 pdfauthor={},
 pdftitle={Crossover in ROPER 2},
 pdfkeywords={},
 pdfsubject={},
 pdfcreator={Emacs 26.1 (Org mode 9.1.14)}, 
 pdflang={English}}
\begin{document}

\maketitle
\tableofcontents


\section{Homologous Crossover in ROPER}
\label{sec:org5967423}
The idea with the crossover mask (or, as its called in the source code, the
"\texttt{xbits}") mechanism is this: With each genome we associate a bitmask (for now,
this is in the form of an unsigned, 64-bit integer, but this is a bit of a
shortcut -- we'd probably like to let it be as long as the longest possible
genome). For the first generation, this value is initialized randomly. During
crossover, the sites of genetic exchange are determined by \emph{combining} (\S
\ref{orgfd1d26a}) the parents' crossover masks: crossover may occur (only) at
those sites (mod 64) where the combined mask has a 1. Every subsequent generation
will inherit its own crossover masks from another \emph{combination} of its parents'
masks -- though, as we'll see, there are good reasons for selecting distinct
combination methods, here: the operation by which the parental masks are combined
to generate the map of potential crossover sites should not necessarily be the
same operation that generates the masks that the offspring will bring to their
own future crossovers.

\lstset{language=rust,label=orgf872d64,caption= ,captionpos=b,numbers=none}
\begin{lstlisting}
pub fn homologous_crossover<R>(mother: &Creature,
                               father: &Creature,
                               mut rng: &mut R) -> Vec<Creature>
where R: Rng, {
    let crossover_degree = *CROSSOVER_DEGREE;
    let bound = usize::min(mother.genome.alleles.len(), 
                           father.genome.alleles.len());
    let xbits = combine_xbits(mother.genome.xbits, 
                              father.genome.xbits, 
                              *CROSSOVER_MASK_COMBINER, rng);
    let child_xbits = combine_xbits(mother.genome.xbits, 
                                    father.genome.xbits, 
                                    *CROSSOVER_MASK_INHERITANCE, rng);
    let sites = xbits_sites(xbits,
                            bound, 
                            crossover_degree, 
                            &mut rng,
    );
    let mut offspring = Vec::new();
    let parents = vec![mother, father];
    let mut i = 0;
    /* Like any respectable couple, the mother and father take
     * turns inseminating one another...
     */
    while offspring.len() < 2 {
        let p0: &Creature = parents[i % 2];
        let p1: &Creature = parents[(i + 1) % 2];
        i += 1;
        let mut egg = p0.genome.alleles.clone();
        let sem = &p1.genome.alleles;
        for site in sites.iter() {
            let codon =
              /* only codons from the father are mutated, but
                since the gender of the parent is decided, each time,
                by chance, this isn't a limitation.
             */
                if rng.gen::<f32>() < *POINTWISE_MUTATION_RATE {
                    mutate_arithmetic(&sem[*site], &mut rng)
                } else {
                    sem[*site]
                };
            egg[*site] = codon;
        }
        let child_gen =
            usize::max(p0.genome.generation, p1.genome.generation) + 1;
        let zygote = Chain {
            alleles: egg,
            metadata: Metadata::new(),
            xbits: random_bit_flip(child_xbits, &mut rng),
            generation: child_gen,
        };
        /* The index will be filled in later, prior to filling
         * the graves of the fallen
         */
        if zygote.entry() != None { /* screen out the gadgetless */
            offspring.push(Creature::new(zygote, 0));
        };
        /*
        if cfg!(debug_assertions) {
            println!("WITH XBITS {:064b}, SITES: {:?}, MATED\n{}\nAND\n{}\nPRODUCING\n{}",
                     xbits,
                     &sites.iter().map(|x| x % bound).collect::<Vec<usize>>(),
                     p0, p1, &offspring[offspring.len()-1]);
            println!("************************************************************");
        }
      */
    }
    offspring
}
\end{lstlisting}

\subsection{Combining the Crossover Masks}
\label{sec:org0518183}
\label{orgfd1d26a}

The precise means of combining the \texttt{xbit} vectors is left open to
experimentation, but a good starting point seems to be bitwise conjunction (the
\texttt{\&} operator). This captures the intuition of restricting crossover to
"genetically compatible" loci -- loci at which the respective \texttt{xbits} of the
parents coincide. Of course, in the beginning, the high bits in the conjunction
of the two masks means nothing at all, and has no real relation to genetic
compatibility. It's just a scaffolding that, it seems, should have the capacity
to \emph{support} emergent compatibility patterns, or a sort of rudimentary
speciation.

But for speciation to occur, the crossover masks should, themselves, be
heritable. The masks of the two parents should be combined into a third,
which will become the child's. This requires a second combination operator,
with the same signature as the first, but we should guard against the
temptation to use the same operator for both. It's pretty clear that \texttt{\&}
is poorly suited to play the role of an inheritance operator: within a
few generations, the crossover masks would converge to 0s. We could
experiment with other canonical boolean operators -- \texttt{xor}, for example,
or \texttt{nand} -- that don't exhibit the fixed-point behaviour that \texttt{and} and
\texttt{or} do, but the most natural choice might just be to use a secondary
\emph{crossover} operation to propagate the masks down the germ lines.
One-point crossover seems like a poor fit, since it would disproportionately
favour the first parent, but a simple, uniform crossover seems well
suited to this task.

In addition, a slow and gentle mutation tendency should probably be 
incorporated as well: the crossover mask that the child will inherit,
and share with its siblings, will be a uniform crossover of its
parents', occasionally perturbed by a single bit-flip mutation.

\lstset{language=rust,label=org0e5a616,caption= ,captionpos=b,numbers=none}
\begin{lstlisting}
/// One-point crossover, between two u64s, as bitvectors.
fn onept_bits<R: Rng>(a: u64, b: u64, rng: &mut R) -> u64 {
    let i = rng.gen::<u64>() % 64;
    let mut mask = ((!0) >> i) << i;
    if rng.gen::<bool>() {
        mask ^= !0
    };
    (mask & a) | (!mask & b)
}

/// Uniform crossover between two u64s, as bitvectors.
fn uniform_bits<R: Rng>(a: u64, b: u64, rng: &mut R) -> u64 {
    let mask = rng.gen::<u64>();
    (mask & a) | (!mask & b)
}

/// A simple mutation operator to use on the crossover mask,
/// prior to passing it on to the offspring.
fn random_bit_flip<R: Rng>(u: u64, rng: &mut R) -> u64 {
  if rng.gen::<f32>() < *CROSSOVER_MASK_MUT_RATE {
      u ^ (1u64 << (rng.gen::<u64>() % 64)) 
  } else {
      u
  }
}

fn combine_xbits<R: Rng>(m_bits: u64,
                         p_bits: u64,
                         combiner: MaskOp,
                         mut rng: &mut R) -> u64 {
    match combiner {
        MaskOp::Xor => m_bits ^ p_bits,
        MaskOp::Nand => !(m_bits & p_bits),
        MaskOp::OnePt => onept_bits(m_bits, p_bits, &mut rng),
        MaskOp::Uniform => uniform_bits(m_bits, p_bits, &mut rng),
        MaskOp::And => m_bits & p_bits,
        MaskOp::Or => m_bits | p_bits,
    }
}
\end{lstlisting}

Once the two parents' \texttt{xbits} have been combined into a crossover mask, we
can use it to generate a list of sites to be used in the genomic crossover
operation. 

\lstset{language=rust,label=org4f4cb75,caption= ,captionpos=b,numbers=none}
\begin{lstlisting}
fn xbits_sites<R: Rng>(
    xbits: u64,
    bound: usize,
    crossover_degree: f32,
    mut rng: &mut R,
) -> Vec<usize> {
    let mut potential_sites = (0..bound)
        .filter(|x| (1u64.rotate_left(*x as u32) & xbits != 0) == *CROSSOVER_XBIT)
        .collect::<Vec<usize>>();
    potential_sites.sort();
    potential_sites.dedup();
    let num = (potential_sites.len() as f32 * crossover_degree).ceil() as usize;

    let mut actual_sites = rand::seq::sample_iter(&mut rng,
                                                  potential_sites.into_iter(), 
                                                  num).unwrap();
  /*
    if cfg!(debug_assertions) {
        println!("{:064b}: potential sites: {:?}", xbits, &potential_sites);
    }

    if cfg!(debug_assertions) {
        println!("actual sites: {:?}", &actual_sites);
    }
     */
    actual_sites
}
\end{lstlisting}

\section{Mutation}
\label{sec:org1cf3bc2}

Without mutation, crossover-driven evolution will eventually stagnate. 

It is desirable that our mutation operators should share a certain minmal
algebraic structure. Each should have an inverse:
\[
(\forall M\in S)(\forall x)(\exists y) M(x) = y \Rightarrow 
(\exists M'\in S) M'(y) = x
\]
and an identity:
\[
(\forall M\in S)(\exists x) M(x) = x
\]
What this means is that over each set of mutation operators -- and
therefore over their union -- the concatenation or succession their
application should form a cyclic group.

In practical terms, this is a generally beneficial property for genetic
operators to possess: all else being equal, they should be designed with
an eye towards neutrality with respect to an arbitrary choice of fitness
functions. By ensuring that the mutation operators are involutive, or, more
generally, that they form a cyclic concatenation group, involution just being
the smallest nontrivial form of such a structure, with a cycle of two, we
(imperfectly) guard against a situation where they ratchet the population into a
small corner of the genotypic (and, consequently, the phenotypic) landscape,
\emph{independent of the fitness function}. (Identity is less significant, in this
context, and is introduced into the mutation operators only
as a way of ensuring closure.) Ratcheting occurs when the genetic
operators are not properly balanced. In the situation where the algebra defined
by concatenation over the mutation operators does \emph{not} form a cyclic group --
when there is "no way back" from some mutation \(M\) by any succession of further
mutations -- ratcheting is inevitable. This problem is distinct from, but
related to, the problem of genetic drift, which it exacerbates. Involutive pairs
of operators, selected with equal probability, provide some safeguard against
this. The \emph{ideal}, in some sense, would be to select genetic operators that
would engender an \emph{ergodic} system under a null fitness function:\footnote{Thanks to
\aimention{Andrea Shepard} Andrea Shepard for this insight.} one whose behaviour is evenly distributed
over the probability landscape it inhabits. In practice, even with fitness
anulled, evolutionary systems rarely exibit such regularity, which has some very
interesting effects on the paths that evolution pursues.

\lstset{language=rust,label=org32c675e,caption= ,captionpos=b,numbers=none}
\begin{lstlisting}
fn mutate_arithmetic <R: Rng> (allele: &Allele, rng: &mut R) -> Allele {
  /* start basic, add more options later */
  let delta = (rng.gen::<isize>() % 16);
  //println!("[+] mutate_arithmetic: delta = {}", delta);
  allele.add(delta)
}
\end{lstlisting}

\section{Dependencies}
\label{sec:orgdcc5585}

For this to work, we'll need just a handful of dependencies: the pseudo-random
number generator library in the \texttt{rand} crate, ROPER's own genotype structs in
\texttt{gen::genotype} (along with the phenotype structures, for inessential reasons:
it seems simpler for now to pass data in its phenome-wrapped state, but this is
a trivial implementation decision, and may change), and a few static parameter
values that we essentially treat as immutable globals in this project, for
the sake of convenience, sparing ourselves quite a bit of parameter clutter.

\lstset{language=rust,label=orgdcbd749,caption= ,captionpos=b,numbers=none}
\begin{lstlisting}
extern crate rand;
use self::rand::{Rng};
use gen::*;
use par::statics::*;
\end{lstlisting}


\section{Putting it Together}
\label{sec:org3a04579}
\lstset{language=rust,label=org325782f,caption= ,captionpos=b,numbers=none}
\begin{lstlisting}
<<crossover-module-dependencies>>
<<mutation>>
<<combining crossover masks>>
<<mapping masks to crossover sites>>
<<homologous crossover>>
\end{lstlisting}
\end{document}